### PR TITLE
Fixed generic inheritance dispatch

### LINKED ISSUE #18845: Object can be implicitly converted into a different base than what is defined
### Example
```nim
type
  Union[T, U] = object of RootObj

  U = object of Union[int, float]

proc test(u: Union[string, RootObj]) = discard

test(U())
```

### Current Output
Compiles and run without error

### Expected Output
Should be a compile-time error

### Additional Information

- Seems to stem from the fact that it is possible to convert between `Union[int, float]` and `Union[string, RootObj]`. If there is a dummy field in `Union` that makes use of the generic parameter, then the implicit conversion will fail. Though I would argue that it should not be possible at all (without `cast`) to convert between generics with different generic parameters, regardless of whether the parameters is used.

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-09-12
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 5d1608c9764416e74b36f6a772f16a66d58ddc28
active boot switches: -d:release -d:nimUseLinenoise
```

### PR DESCRIPTION
**Pull request by beef331**
Closes: https://github.com/nim-lang/Nim/issues/18845

### CODE REVIEW

File: compiler/sigmatch.nim [hunk 1]
```diff
@@ -1412,6 +1412,11 @@ proc typeRel(c: var TCandidate, f, aOrig: PType,
     let origF = f
     var f = if prev == nil: f else: prev
 
+    if a.len > 0:
+      let skippedA = a.skipTypesOrNil({tyObject}) # Skips if it's inheriting from a generic
+      if skippedA != nil:
+        a = skippedA
+
```
Comment (Araq) [inline compiler/sigmatch.nim:1419]: I don't understand this patch, looks dubious.

Comment (beef331) [reply compiler/sigmatch.nim:1419]: I think you might be right but I cannot think of an edge case to prove it. The only time the skippedA applies is on `object of X`, or atleast the only time I think of. I'll make more elaborate tests(more just to try to break it than anything).

Comment (Araq) [reply compiler/sigmatch.nim:1419]: There is not a single place in the compiler where we ever skip `tyObject`, what does that do? Skip to what?
